hal/src/thumbv7em/nvm.rs

```diff
@@ -0,0 +1,573 @@
+
+        for i in 0..6 {
+            buffer |=
+                unsafe { core::ptr::read_volatile(base_addr.offset(i as isize)) as u64 } << (i * 8);
+        }
+
+        CalibrationArea(buffer)
+    }
+
+    /// Read the calibration area for temperatures
+    #[inline]
+    pub fn temperatures_calibration_area(&self) -> TemperaturesCalibrationArea {
+        let mut buffer = 0_u128;
+        let base_addr: *const u8 = 0x0080_0100 as *const u8;
+
+        for i in 0..11 {
+            buffer |= unsafe { core::ptr::read_volatile(base_addr.offset(i as isize)) as u128 }
+                << (i * 8);
+        }
+
+        TemperaturesCalibrationArea(buffer)
+    }
+
+    /// Enable/disable boot protection on/off
+    ///
+    /// Userpage's NVM BOOT field defines a memory region that is supposed to be
+    /// protected. `NVMCTRL.STATUS.BOOTPROT` is a readonly HW register populated
+    /// on reset with a value from a userpage. By default, 0
+    #[inline]
+    pub fn boot_protection(&mut self, protect: bool) -> Result<()> {
+        // Check if requested state differs from current state
+        if self.is_boot_protected() != protect {
+            // Wait until ready
+            while !self.is_ready() {}
+
+            // Requires both command and key so the command is allowed to execute
+            if !protect {
+                // Issue Set boot protection disable (disable bootprotection)
+                self.command_sync(CMD_AW::SBPDIS);
+            } else {
+                // Issue Clear boot protection disable (enable bootprotection)
+                self.command_sync(CMD_AW::CBPDIS);
+            }
+
+            self.manage_error_states()
+        } else {
+            Err(Error::NoChangeBootProtection)
+        }
+    }
+
+    /// Write to flash memory from a slice
+    ///
+    /// If `destination_address` is not word-aligned, an error is returned.
+    #[inline]
+    pub unsafe fn write_from_slice(
+        &mut self,
+        destination_address: u32,
+        source_slice: &[u32],
+    ) -> Result<()> {
+        let source_address = source_slice.as_ptr() as u32;
+        let words = source_slice.len() as u32;
+        self.write(destination_address, source_address, words)
+    }
+
+    /// Write to flash memory
+    ///
+    /// If either `destination_address` or `source_address` are not
+    /// word-aligned, an error is returned.
+    #[inline]
+    pub unsafe fn write(
+        &mut self,
+        destination_address: u32,
+        source_address: u32,
+        words: u32,
+    ) -> Result<()> {
+        // Length of memory step
+        let step_size: u32 = core::mem::size_of::<u32>() as u32;
+        // Length of data in bytes
+        let length = words * step_size;
+
+        let read_addresses = source_address..(source_address + length);
+        let write_addresses = destination_address..(destination_address + length);
+
+        if source_address % step_size != 0 {
+            return Err(Error::Alignment);
+        }
```

Comment:
Is a slice guaranteed to be 4-aligned then :thinking:

Reply:
Or maybe
```rust
if source_address % step_size != 0 {
    return Err(Error::Alignment);
}
```
is redundant

Reply:
Or maybe
```rust
if source_address % step_size != 0 {
    return Err(Error::Alignment);
}
```
is just redundant

Reply:
this function itself is just getting any ole `u32`, so doing this extra step seems like a good idea.

Reply:
Ok then, I guess we can merge that in. If we find any issues along the way we can patch them later in separate PRs.